README


This archive contains the distribution Test-Deep-NumberTolerant,
version 0.001:

  A Test::Deep plugin for testing numbers within a tolerance range

This software is copyright (c) 2014 by Karen Etheridge.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v5.021.



Changes
Revision history for Test-Deep-NumberTolerant

0.001     2014-10-28 01:07:12Z
          - Initial release.


lib/Test/Deep/NumberTolerant.pm
use strict;
use warnings;
package Test::Deep::NumberTolerant;
# git description: 367b0d1
$Test::Deep::NumberTolerant::VERSION = '0.001';
# ABSTRACT: A Test::Deep plugin for testing numbers within a tolerance range
# KEYWORDS: testing tests plugin numbers tolerance range epsilon uncertainty
# vim: set ts=8 sw=4 tw=78 et :

use parent 'Test::Deep::Cmp';
use Number::Tolerant;
use namespace::clean;
use Exporter 'import';

our @EXPORT = qw(within_tolerance);

sub within_tolerance
{
    my ($number, @tolerance_args) = @_;
    return __PACKAGE__->new($number, @tolerance_args);
}

sub init
{
    my $self = shift;
    $self->{tolerance} = tolerance(@_);
}

sub descend
{
    my ($self, $got) = @_;
    return $got == $self->{tolerance};
}

sub diag_message
{
    my ($self, $where) = @_;

    return 'Checking ' . $where . ' against ' . $self->{tolerance};
}

# we do not define a diagnostics sub, so we get the one produced by deep_diag
# showing exactly what part of the data structure failed. This calls renderGot
# and renderVal:

sub renderGot
{
    my ($self, $got) = @_;
    return defined $self->{error_message}
        ? $self->{error_message}
        : 'failed';  # TODO?  $got . ' is not ' . $self->{tolerance};
}

sub renderExp
{
    my $self = shift;
    return 'no error';
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Test::Deep::NumberTolerant - A Test::Deep plugin for testing numbers within a tolerance range

=head1 VERSION

version 0.001

=head1 SYNOPSIS

    use Test::More;
    use Test::Deep;
    use Test::Deep::NumberTolerant;

    cmp_deeply(
        {
            counter => 123,
        },
        {
            counter => within_tolerance(100, plus_or_minus => 50),
        },
        'counter field is 100 plus or minus 50',
    );

=head1 DESCRIPTION

C<Test::Deep::NumberTolerant> provides the sub C<within_tolerance> to indicate
that the data being tested matches the equivalent C<tolerance(...)> value.

I wrote this because I found myself doing this a lot:

    cmp_deeply(
        $thing,
        methods(
            delete_time => methods(epoch => code( sub { $_[0] == tolerance(time(), plus_or_minus => 2) || (0, "got $_[0], expected ", time()) } )),
        ),
        'object has been deleted',
    );

With this module, this can be simplified to the much more readable:

    cmp_deeply(
        $thing,
        methods(
            delete_time => methods(epoch => within_tolerance(time(), plus_or_minus => 2)),
        ),
        'object has been deleted',
    );

=head1 FUNCTIONS

=head2 C<within_tolerance>

Exported by default; to be used within a L<Test::Deep> comparison function
such as L<cmp_deeply|Test::Deep/COMPARISON FUNCTIONS>.  Accepted arguments are
the same as for C<L<Number::Tolerant/tolerance>>.

=for Pod::Coverage descend
diag_message
init
renderExp
renderGot

=head1 SUPPORT

=for stopwords irc

Bugs may be submitted through L<the RT bug tracker|https://rt.cpan.org/Public/Dist/Display.html?Name=Test-Deep-NumberTolerant>
(or L<bug-Test-Deep-NumberTolerant@rt.cpan.org|mailto:bug-Test-Deep-NumberTolerant@rt.cpan.org>).
I am also usually active on irc, as 'ether' at C<irc.perl.org>.

=head1 SEE ALSO

=over 4

=item *

L<Test::Deep>

=item *

L<Number::Tolerant>

=back

=head1 AUTHOR

Karen Etheridge <ether@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Karen Etheridge.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Build.PL
# This Build.PL for Test-Deep-NumberTolerant was generated by
# Dist::Zilla::Plugin::ModuleBuildTiny::Fallback 0.009
use strict;
use warnings;

if (eval 'use Module::Build::Tiny 0.039; 1')
{
    print "Congratulations, your toolchain understands 'configure_requires'!\n\n";

    # This section for Test-Deep-NumberTolerant was generated by Dist::Zilla::Plugin::ModuleBuildTiny 0.007.
    use strict;
    use warnings;

    use 5.006;
    # use Module::Build::Tiny 0.039;
    Build_PL();
}
else
{
    $ENV{PERL_MB_FALLBACK_SILENCE_WARNING} or warn <<'EOW';
*** WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING ***

If you're seeing this warning, your toolchain is really, really old* and you'll
almost certainly have problems installing CPAN modules from this century. But
never fear, dear user, for we have the technology to fix this!

If you're using CPAN.pm to install things, then you can upgrade it using:

    cpan CPAN

If you're using CPANPLUS to install things, then you can upgrade it using:

    cpanp CPANPLUS

If you're using cpanminus, you shouldn't be seeing this message in the first
place, so please file an issue on github.

This public service announcement was brought to you by the Perl Toolchain
Gang, the irc.perl.org #toolchain IRC channel, and the number 42.

----

* Alternatively, you are running this file manually, in which case you need
to learn to first fulfill all configure requires prerequisites listed in
META.yml or META.json -- or use a cpan client to install this distribution.

You can also silence this warning for future installations by setting the
PERL_MB_FALLBACK_SILENCE_WARNING environment variable, but please don't do
that until you fix your toolchain as described above.

EOW
    sleep 10 if -t STDIN && (-t STDOUT || !(-f STDOUT || -c STDOUT));


    # This section was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.021.
    use strict;
    use warnings;

    require Module::Build; Module::Build->VERSION(0.28);


    my %module_build_args = (
      "build_requires" => {},
      "configure_requires" => {
        "ExtUtils::MakeMaker" => 0,
        "Module::Build::Tiny" => "0.039"
      },
      "dist_abstract" => "A Test::Deep plugin for testing numbers within a tolerance range",
      "dist_author" => [
        "Karen Etheridge <ether\@cpan.org>"
      ],
      "dist_name" => "Test-Deep-NumberTolerant",
      "dist_version" => "0.001",
      "license" => "perl",
      "module_name" => "Test::Deep::NumberTolerant",
      "recommends" => {},
      "recursive_test_files" => 1,
      "requires" => {
        "Exporter" => 0,
        "Number::Tolerant" => 0,
        "Test::Deep::Cmp" => 0,
        "namespace::clean" => 0,
        "parent" => 0,
        "perl" => "5.006",
        "strict" => 0,
        "warnings" => 0
      },
      "script_files" => [],
      "test_requires" => {
        "ExtUtils::MakeMaker" => 0,
        "File::Spec" => 0,
        "Test::Deep" => 0,
        "Test::More" => 0,
        "Test::Tester" => "0.108",
        "if" => 0
      }
    );


    my %fallback_build_requires = (
      "ExtUtils::MakeMaker" => 0,
      "File::Spec" => 0,
      "Test::Deep" => 0,
      "Test::More" => 0,
      "Test::Tester" => "0.108",
      "if" => 0
    );


    unless ( eval { Module::Build->VERSION(0.4004) } ) {
      delete $module_build_args{test_requires};
      $module_build_args{build_requires} = \%fallback_build_requires;
    }

    my $build = Module::Build->new(%module_build_args);

    $build->create_build_script;
}


Makefile.PL
# This Makefile.PL for Test-Deep-NumberTolerant was generated by
# Dist::Zilla::Plugin::MakeMaker::Fallback 0.013
# and Dist::Zilla::Plugin::MakeMaker::Awesome 0.31.
# Don't edit it but the dist.ini and plugins used to construct it.

use strict;
use warnings;

BEGIN {
my %configure_requires = (
    'ExtUtils::MakeMaker' => '0',
    'Module::Build::Tiny' => '0.039',
);

my @missing = grep {
    ! eval "require $_; $_->VERSION($configure_requires{$_}); 1"
} keys %configure_requires;

if (not @missing)
{
    print "Congratulations, your toolchain understands 'configure_requires'!\n\n";
}
else
{
    $ENV{PERL_MM_FALLBACK_SILENCE_WARNING} or warn <<'EOW';
*** WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING ***

If you're seeing this warning, your toolchain is really, really old* and you'll
almost certainly have problems installing CPAN modules from this century. But
never fear, dear user, for we have the technology to fix this!

If you're using CPAN.pm to install things, then you can upgrade it using:

    cpan CPAN

If you're using CPANPLUS to install things, then you can upgrade it using:

    cpanp CPANPLUS

If you're using cpanminus, you shouldn't be seeing this message in the first
place, so please file an issue on github.

If you're installing manually, please retrain your fingers to run Build.PL
when present instead.

This public service announcement was brought to you by the Perl Toolchain
Gang, the irc.perl.org #toolchain IRC channel, and the number 42.

----

* Alternatively, you are doing something overly clever, in which case you
should consider setting the 'prefer_installer' config option in CPAN.pm, or
'prefer_makefile' in CPANPLUS, to 'mb" and '0' respectively.

You can also silence this warning for future installations by setting the
PERL_MM_FALLBACK_SILENCE_WARNING environment variable.

EOW

    sleep 10 if -t STDIN && (-t STDOUT || !(-f STDOUT || -c STDOUT));
}
}

use 5.006;
use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "A Test::Deep plugin for testing numbers within a tolerance range",
  "AUTHOR" => "Karen Etheridge <ether\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "Module::Build::Tiny" => "0.039"
  },
  "DISTNAME" => "Test-Deep-NumberTolerant",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "Test::Deep::NumberTolerant",
  "PREREQ_PM" => {
    "Exporter" => 0,
    "Number::Tolerant" => 0,
    "Test::Deep::Cmp" => 0,
    "namespace::clean" => 0,
    "parent" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::Deep" => 0,
    "Test::More" => 0,
    "Test::Tester" => "0.108",
    "if" => 0
  },
  "VERSION" => "0.001",
  "test" => {
    "TESTS" => "t/*.t"
  }
);

my %FallbackPrereqs = (
  "Exporter" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Module::Build::Tiny" => "0.039",
  "Number::Tolerant" => 0,
  "Test::Deep" => 0,
  "Test::Deep::Cmp" => 0,
  "Test::More" => 0,
  "Test::Tester" => "0.108",
  "if" => 0,
  "namespace::clean" => 0,
  "parent" => 0,
  "strict" => 0,
  "warnings" => 0
);

unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
